### PR TITLE
Python. Homework #1. Task #1 Порсев Егор

### PR DESCRIPTION
**Pull request by Furetur**

### CODE REVIEW

File: hw1/matrices.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from math import sqrt, acos, isclose, pi
+from typing import List
```
Comment (areyde) [inline hw1/matrices.py]: Why is this import here, but never used?

File: hw1/matrices.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def angle_between_vectors(vector1, vector2):
+    return acos(dot_product(vector1, vector2) / (magnitude(vector1) * magnitude(vector2)))
+
+
+def check_and_matrix_shape(matrix):
+    if len(matrix) == 0 or len(matrix[0]) == 0 or any((len(row) != len(matrix[0]) for row in matrix)):
+        raise IndexError("Matrix cannot be empty or contain empty rows")
+    if any((len(row) != len(matrix[0]) for row in matrix)):
```
Comment (areyde) [inline hw1/matrices.py]: In these two checks you are repeating the same assertion. Try to write the first one a bit differently, so that you don't check for the same thing twice.

File: hw1/matrices.py (hunk 1 above)
Comment (areyde) [reply hw1/matrices.py]: But hey, since this is here, let's actually add Typing to all the functions.

File: hw1/matrices.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    assert dot_product(orthogonal1, orthogonal2) == 0
+    assert isclose(magnitude([1 / sqrt(2), 1 / sqrt(2)]), 1)
+    assert isclose(angle_between_vectors([1, -1, 1], [1, 1, 0]), pi / 2)
+    assert matrix_transpose([[1], [2], [3]]) == [[1, 2, 3]]
+    assert matrix_sum([[1, 2, 3], [-1, -2, -3]], [[-1, -2, -3], [1, 2, 3]]) == [[0, 0, 0], [0, 0, 0]]
+    assert matrix_multiplication([[1, 2, 3], [4, 5, 6], [7, 8, 9]], [[1, 0, 0], [0, 1, 0], [0, 0, 1]]) == [
+        [1, 2, 3],
+        [4, 5, 6],
+        [7, 8, 9],
+    ]
```
Comment (areyde) [inline hw1/matrices.py]: This last assert is a bit diffcult to read. Can you maybe make the long asserts (this and the previous one) a bit more readable? Maybe use some newlines if Black allows for it?

**Review by areyde — CHANGES_REQUESTED** [review]
You did a great job! Very nice reuse of functions and good checks. Just a couple of very minor fixes to do, will take you just several minutes.

File: hw1/matrices.py (hunk 1 above)
Comment (Furetur) [reply hw1/matrices.py]: It is there because I actually tried adding typings but failed :). Could you please help me a little bit with that by answering a question. What do you think the type of a vector should be? It should be something like `Iterable[Union[int, float]]` but that's too ugly. I couldn't find any other numeral types and defining my own type like `typedef number = int | float` feels wrong :)

Comment (areyde) [reply hw1/matrices.py]: The first variant is actually good. With `Union` and everything. But you are right that it is weird to write it everywhere, it's long and unreadable. So try _Type aliases_ https://docs.python.org/3/library/typing.html and just create a neat alias for this type.

Comment (areyde) [reply hw1/matrices.py]: Also, I found this. https://www.python.org/dev/peps/pep-0484/#the-numeric-tower

You don't need to write Union[int, float], you can just write float, then int will be autmoatically accepted.

File: hw1/matrices.py (hunk 3 above)
Comment (Furetur) [reply hw1/matrices.py]: Black actually didn't allow any newlines anywhere near this line so I killed it and the problem is now solved.

**Review by areyde — APPROVED** [review]
Good job. The last assert still looks vrey rough, but since tests weren't obligatory for the first homework, it's alright. 

The problem is not Black, the problem is writing it in a sing expression. For the future, I suggest you to use `unittest` testcases, and first write the necessary objects (nicely) and only then compare.

For example, this is how your coursemate did it: https://github.com/GirZ0n/SPBU-Homework-4/blob/main/tests/homework1/task1/test_matrix.py#L363
